Update Testing documentation (#1054)
vapor/docs: docs/advanced/testing.md

# Testing

## VaporTesting

Vapor includes a module named `VaporTesting` that provides test helpers built on `Swift Testing`. These testing helpers allow you to send test requests to your Vapor application programmatically or running over an HTTP server.

> !!! note
>     For newer projects or teams adopting Swift concurrency, `VaporTesting` is highly recommended due to its simplicity and integration with Vapor.

Reviewer: Not sure what exactly would be the best but I'd want to sneak in a mention of "XCT" in there just so it's clear what it's recommended over. Maybe something like this but not sure:
```suggestion
!!! note
    For newer projects or teams adopting Swift concurrency, `Swift Testing` is highly recommended over `XCT`.
```

Author: I like the suggestion. I don't know what's better, to say Swift Testing or VaporTesting. And the other part I'll name it XCTest or XCTVapor, instead of XCT. What do you think? cc/ @0xTim

Reviewer 2: `XCT` doesn't really mean anything, so for this instance we should say `XCTest`

```swift
let package = Package(
    ...
    dependencies: [
        .package(url: "https://github.com/vapor/vapor.git", from: "4.110.1")
    ],
    targets: [
        ...
        .testTarget(name: "AppTests", dependencies: [
            .target(name: "App"),
            .product(name: "VaporTesting", package: "vapor"),
        ])
    ]
)
```

!!! warning
    Be sure to use the corresponding testing module, as failing to do so can result in Vapor test failures not being properly reported.

Then, add `import VaporTesting` and `import Testing` at the top of your test files. Create structs with a `@Suite` name to write test cases. 

```swift
@testable import App
import VaporTesting
import Testing

@Suite("App Tests")
struct AppTests {
    @Test("Test Stub")
    func stub() async throws {
    	// Test here.
    }
}
```

Each function marked with `@Test` will run automatically when your app is tested.

To ensure your tests run in a serialized manner (e.g., when testing with a database), include the `.serialized` option in the test suite declaration:

```swift
@Suite("App Tests with DB", .serialized)
```

> #### Running Tests

Reviewer 2: I don't think we really need this, open to keeping it though. The doc isn't about how to run tests, it's how to use Vapor Testing

Author: Yeah, I kept it because it was already there, but I'm happy to remove this.

> Use `cmd+u` with the `-Package` scheme selected to run tests in Xcode. Use `swift test --enable-test-discovery` to test via the CLI.

Reviewer 2: Test discovery is no longer a thing

Author: Glad to hear this. I'll remove this.

    do {
        try await configure(app)
        try await app.autoMigrate()
        try await test(app)
        try await app.autoRevert()   
    }
    catch {
        try? await app.autoRevert()
        try await app.asyncShutdown()
        throw error
    }
    try await app.asyncShutdown()
}
```

## Using XCTest

If you prefer XCTest over the Swift Testing framework, or if a project requires compatibility with a more traditional approach, XCTest is fully supported. Here’s how you can get started:

First ensure that the `XCTVapor` module has been added to your package's test target.

```swift
let package = Package(
    ...
    dependencies: [
        .package(url: "https://github.com/vapor/vapor.git", from: "4.0.0")
    ],
    targets: [
        ...
        .testTarget(name: "AppTests", dependencies: [
            .target(name: "App"),
            .product(name: "XCTVapor", package: "vapor"),
        ])
    ]
)
```

Then, add `import XCTVapor` at the top of your test files. Create classes extending `XCTestCase` to write tests cases.

```swift
import XCTVapor

final class MyTests: XCTestCase {
    func testStub() throws {
        // Test here.

>     }

Reviewer 2: Can we just copy over the old docs? So basically what we had, plus the note, about preferring VaporTesting (but add it at the top) with the new Swift Testing blurb at the top?

Author: Sure! Let me update the PR.
